angular/projects/spark-angular/src/lib/components/sprk-autocomplete/sprk-autocomplete.component.ts

```diff
@@ -0,0 +1,395 @@
+          this.input.ref.nativeElement.parentNode,
+          'aria-expanded',
+          'false',
+        );
+
+        // Remove aria-activedescendant from the input
+        this.renderer.removeAttribute(
+          this.input.ref.nativeElement,
+          'aria-activedescendant',
+        );
+      }
+
+      // remove aria-selected and highlight class from all list items
+      this.removeAllHighlights();
+
+      this.closedEvent.emit();
+      this.isOpen = false;
+    }
+  }
+
+  /**
+   * Show the Autocomplete results list.
+   */
+  showResults(): void {
+    if (this.results) {
+      this.renderer.removeClass(
+        this.results.nativeElement,
+        'sprk-c-Autocomplete__results--hidden',
+      );
+
+      if (this.input) {
+        // Set aria-expanded on the input's parent
+        this.renderer.setAttribute(
+          this.input.ref.nativeElement.parentNode,
+          'aria-expanded',
+          'true',
+        );
+      }
+
+      this.openedEvent.emit();
+      this.isOpen = true;
+    }
+  }
+
+  /**
+   * @ignore
+   * Track the index of the currently-highlighted Autocomplete result item.
+   */
+  highlightedIndex = -1;
+
+  /**
+   * @ignore
+   */
+  ngAfterContentInit(): void {
+    if (this.results) {
+      if (this.isOpen) {
+        this.showResults();
+      } else {
+        this.hideResults();
+      }
+    }
+
+    // if itemSelectedEvent is specified, also set that as the click event on each result
+    if (this.itemSelectedEvent) {
+      this.resultItems.forEach((element) => {
+        element.clickedEvent = this.itemSelectedEvent;
+      });
+    }
+  }
+
+  /**
+   * @ignore
+   * This logic needs to happen in this event so that the correct DOM
+   * attributes will have been set.
+   */
+  ngAfterViewInit(): void {
+    if (this.input) {
+      this.calculateResultsWidth();
+    }
+
+    if (this.results) {
+      if (this.input) {
+        // set aria-controls on the input to the id of the results
+        this.generateAriaControls(
+          this.input.ref.nativeElement,
+          this.results.nativeElement,
+        );
+
+        // set aria-owns on the input's parent to the id of the results
+        this.generateAriaOwns(
+          this.input.ref.nativeElement.parentNode,
+          this.results.nativeElement,
+        );
+      }
+    }
+  }
+
+  /**
+   * @ignore
+   * Set the max-width on the Autocomplete results equal to the current
+   * width of the input element.
+   */
+  calculateResultsWidth() {
+    if (this.input) {
+      const currentInputWidth = this.input.ref.nativeElement.offsetWidth;
+
+      if (this.results) {
+        this.renderer.setAttribute(
+          this.results.nativeElement,
+          'style',
+          'max-width:' + currentInputWidth + 'px',
+        );
+      }
+    }
+  }
+  /** @ignore */
+  isUpPressed = (e) => e.key === 'ArrowUp' || e.keyCode === 38;
+  /** @ignore */
+  isDownPressed = (e) => e.key === 'ArrowDown' || e.keyCode === 40;
```

Comment:
These would great to add as utilities. We would use them in tabs and in modal. Could you add them as utilities? I made a utils folder in angular in my masthead pr with some I had made. We previously did not have a utils folder.

Reply:
Yes definitely. I'll also move `generateAriaOwns` and `generateAriaControls` to utilities.

Reply:
update, we are not moving these two to be utils at this time.